frontend_app_admin_workspace_user/test/container/AdminWorkspaceUser.spec.js

```diff
@@ -0,0 +1,240 @@
+import React from 'react'
+import { expect } from 'chai'
+import { shallow } from 'enzyme'
+import {
+  mockGetWorkspaces200,
+  mockGetWorkspaceMembers200,
+  mockGetUsers200,
+  mockGetUserDetails200
+} from '../apiMock.js'
+import { AdminWorkspaceUser } from '../../src/container/AdminWorkspaceUser.jsx'
+
+describe('<AdminWorkspaceUser />', () => {
+  const props = {
+    i18n: {},
+    registerCustomEventHandlerList: () => { },
+    registerLiveMessageHandlerList: () => { },
+    t: key => key,
+    data: {
+      content: {
+        workspaceList: [],
+        userList: []
+      },
+      config: {
+        type: 'workspace',
+        translation: {},
+        apiUrl: 'http://localhost/api',
+        system: {
+          config: {
+            email_notification_activated: false
+          }
+        }
+      },
+      loggedUser: {
+        lang: 'en',
+        user_id: 1
+      }
+    }
+  }
+
+  describe('TLM handlers', () => {
+    describe('eventType workspace', () => {
+      props.data.config.type = 'workspace'
+      mockGetWorkspaces200(props.data.config.apiUrl, [{ workspace_id: 1, label: 'Hello', description: '' }]).persist()
+      mockGetWorkspaceMembers200(props.data.config.apiUrl, 1, []).persist()
+
+      describe('handleWorkspaceCreated', () => {
+
+        const workspace = {
+          workspace_id: 5,
+          label: 'A workspace',
+          description: ''
+        }
+        mockGetWorkspaceMembers200(props.data.config.apiUrl, workspace.workspace_id, [])
+        const wrapper = shallow(<AdminWorkspaceUser {...props} />)
+
+        it('should add the created workspace to the end of the list', async () => {
+          const tlmData = {
+            workspace: workspace
+          }
+          const previousWorkspaceListLength = wrapper.state('content').workspaceList.length
+          await wrapper.instance().handleWorkspaceCreated(tlmData)
```

Comment:
In order to have test files which are consistent in their structure, you should put the call of the function tested in a `describe` and each assertions should be located in different `it`. `frontend/test/src/container/Dashboard.spec.js` is a good example

Reply:
OK, but when I make this change, it breaks most tests.

Working version:

```javascript
        const wrapper = shallow(<AdminWorkspaceUser {...props} />)
        it('should add the created workspace to the end of the list', () => {
          wrapper.instance().handleWorkspaceCreated(tlmData)
          const workspaceList = wrapper.state('content').workspaceList
          const lastWorkspace = workspaceList[workspaceList.length - 1]
          expect(lastWorkspace).to.deep.equal({ ...tlmData.workspace, memberList: [] })
        })
```

Failing version :

```javascript
        const wrapper = shallow(<AdminWorkspaceUser {...props} />)
        wrapper.instance().handleWorkspaceCreated(tlmData)
        it('should add the created workspace to the end of the list', () => {
          const workspaceList = wrapper.state('content').workspaceList
          const lastWorkspace = workspaceList[workspaceList.length - 1]
          expect(lastWorkspace).to.deep.equal({ ...tlmData.workspace, memberList: [] })
        })
```

I must miss something but I don't see what… @metzgegu Do you have an idea?

Reply:
Ok, putting the `handleWorkspaceCreated` in a `before` hook works… but I'm not a fan.